### ISSUE TITLE
Value types with implicit string cast

### ISSUE THREAD

**Issue by codespare** [microsoft/RoslynClrHeapAllocationAnalyzer#19]
The analyzer (v1.0 VS 2015 extension) squiggles for value types even when an overloaded call involving an implicit string conversion operator is selected by the compiler, thus avoiding boxing.
An example is Console.WriteLine that has both (object) and (string) overloads, the later being selected for value types having an implicit string conversion operator from what I can tell.
I'd say that's an edge case at best. Your extension is brilliant and makes for a fun game playing against the compiler, thank you :)

**Comment by mjsabby** [comment]
@codespare -- if you can attach a quick unit test or sample, I'm going to fix it. I understand the issue, it is one of those things that'll need a little bit of thought to not add a special case.

**Comment by codespare** [comment]
Sure, I'll put something together today or tomorrow. I noticed it chasing squiggles with the msiler extension, but in a somewhat messy code base, so that needs some cleanup.

**Comment by codespare** [comment]
A hack of a fix could be:
``` c#
                var argumentTypeInfo = semanticModel.GetTypeInfo(argument.Expression, cancellationToken);
                var conversionInfo = semanticModel.GetConversion(argument.Expression, cancellationToken);
                if (argumentTypeInfo.Type?.IsValueType == true && argumentTypeInfo.ConvertedType?.IsValueType == false && !(conversionInfo.IsImplicit && conversionInfo.IsUserDefined)) {
                    CheckTypeConversion(argumentTypeInfo, reportDiagnostic, argument.Expression.GetLocation(), filePath);
                }
```
in TypeConversionAllocationAnalyzer.ArgumentSyntaxCheck, although I wouldn't bet the farm on it yet :)

**Comment by edespong** [comment]
It seems like the proposed fix will work. I think, however, that the issue here is more general and needs a fix in the `CheckTypeConversion` method. For example, we get the exact same issue (i.e. a warning when there should be none) if doing a yield return:

```
public struct AStruct {
  public IEnumerator<string> GetEnumerator() {
    yield return this;
  }
  public static implicit operator string(AStruct astruct) {
   return "";
  }
}
```

It seems to me that `GetConversion `is the way to go though because there we get access to the `IsBoxing `property, which is exactly what we are trying to model with the current `typeInfo.Type?.IsValueType == true && typeInfo.ConvertedType?.IsValueType != true`.

**Comment by cezarypiatek** [comment]
@mjsabby It was fixed by #34, could you close it?
